src/dmd/dmodule.d

```diff
@@ -1355,3 +1382,221 @@ struct ModuleDeclaration
         return buf.extractString();
     }
 }
+
+struct ModuleComponentRange
+{
+    Identifiers* packages;
+    Identifier name;
+    size_t index;
+    @property auto totalLength() const { return packages.dim + 1; }
+
+    @property auto empty() { return index > packages.dim; }
+    @property auto front() const { return (index == packages.dim) ? name : (*packages)[index]; }
+    void popFront() { index++; }
+}
+
+// returns whether the given module should be included in the compilation
+private bool includeImportedModuleCheck(ModuleComponentRange components)
+    in { assert(global.params.includeImports); } body
+{
+    createMatchNodes();
+    {
+        size_t nodeIndex = 0;
+        for (; nodeIndex < matchNodes.dim;)
+        {
+            //printf("matcher ");printMatcher(nodeIndex);printf("\n");
+            auto info = matchNodes[nodeIndex++];
+            if (info.depth <= components.totalLength())
+            {
+                size_t nodeOffset = 0;
+                foreach (component; components)
+                {
+                    if (nodeOffset >= info.depth)
+                        break;
+                    if (!component.equals(matchNodes[nodeIndex + nodeOffset].id))
+                    {
+                        goto Lmismatch;
+                    }
+                    nodeOffset++;
+                }
+                // MATCH
+                //printf("matcher ");printMatcher(nodeIndex - 1);
+                //printf(" MATCHES module '");components.print();printf("'\n");
+                return !info.isExclude;
+            Lmismatch:
+                //printf("matcher ");printMatcher(nodeIndex-1);
+                //printf(" does not match module '");components.print();printf("'\n");
+            }
+            nodeIndex += info.depth;
+        }
+        assert(nodeIndex == matchNodes.dim, "code bug");
+    }
+    return includeByDefault;
+}
+
+// Matching module names is done with an array of matcher nodes.
+// The nodes are sorted by "component depth" from largest to smallest
+// so that the first match is always the longest (best) match.
+private struct MatcherNode
+{
+    union
+    {
+        struct
+        {
+            ushort depth;
+            bool isExclude;
+        }
+        Identifier id;
+    }
```

Comment:
unclear how discrimination is done - care to assert about e.g. the size of Identifier?

Reply:
I guess by "discrimination" you mean, how does the code know if it's looking at an identifier node or an "depth/info" node?  The node type is determined by the ordering, each "match sequence" is started by a "depth/info" node which indicates how many identifiers it has (by it's depth).  So the structure looks something like this:
```
| matchSequence0 (depthForSequence0) | id(0) | id(1) | ... | id(depthForSequence0-1) |
| matchSequence1 (depthForSequence1) | id(0) | id(1) | ... | id(depthForSequence1-1) |
...
```